pkg/zone/available_services.go

```diff
@@ -0,0 +1,149 @@
+		return nil, err
+	}
+	return &ZoneAvailableServicesTracker{
+		logger:            logger,
+		metric:            metric,
+		resManager:        resManager,
+		meshCache:         meshCache,
+		interval:          interval,
+		ingressTagFilters: ingressTagFilters,
+		zone:              zone,
+	}, nil
+}
+
+func (t *ZoneAvailableServicesTracker) Start(stop <-chan struct{}) error {
+	t.logger.Info("starting")
+	ticker := time.NewTicker(t.interval)
+	ctx := user.Ctx(context.Background(), user.ControlPlane)
+
+	for {
+		select {
+		case <-ticker.C:
+			start := time.Now()
+			if err := t.updateZoneIngresses(ctx); err != nil {
+				t.logger.Error(err, "couldn't update ZoneIngress available services")
+			}
+			t.metric.Observe(float64(time.Since(start).Milliseconds()))
+		case <-stop:
+			t.logger.Info("stopping")
+			return nil
+		}
+	}
+}
+
+func (t *ZoneAvailableServicesTracker) NeedLeaderElection() bool {
+	return true
+}
+
+func availableServicesEqual(services []*mesh_proto.ZoneIngress_AvailableService, other []*mesh_proto.ZoneIngress_AvailableService) bool {
+	if len(services) != len(other) {
+		return false
+	}
+	for i := range services {
+		if !proto.Equal(services[i], other[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+func (t *ZoneAvailableServicesTracker) getIngressExternalServices(
+	aggregatedMeshCtxs xds_context.AggregatedMeshContexts,
+) []*core_mesh.ExternalServiceResource {
+	var externalServices []*core_mesh.ExternalServiceResource
+
+	for _, mesh := range aggregatedMeshCtxs.Meshes {
+		if !mesh.ZoneEgressEnabled() {
+			continue
+		}
+
+		meshCtx := aggregatedMeshCtxs.MustGetMeshContext(mesh.GetMeta().GetName())
+		// look for external services that are only available in my zone and expose them
+		for _, es := range meshCtx.Resources.ExternalServices().Items {
+			if es.Spec.Tags[mesh_proto.ZoneTag] == t.zone {
+				externalServices = append(externalServices, es)
+			}
+		}
+	}
+
+	return externalServices
+}
+
+func (t *ZoneAvailableServicesTracker) updateZoneIngresses(ctx context.Context) error {
+	aggregatedMeshCtxs, err := xds_context.AggregateMeshContexts(ctx, t.resManager, t.meshCache.GetMeshContext)
+	if err != nil {
+		return err
+	}
+
+	zis := core_mesh.ZoneIngressResourceList{}
+	if err := t.resManager.List(ctx, &zis); err != nil {
+		return err
+	}
+
+	availableServices := ingress.GetAvailableServices(
+		aggregatedMeshCtxs.AllDataplanes(),
+		aggregatedMeshCtxs.AllMeshGateways(),
+		t.getIngressExternalServices(aggregatedMeshCtxs),
+		t.ingressTagFilters,
+	)
+
+	for _, zi := range zis.Items {
+		// Initially zone is empty
+		if zi.Spec.Zone != "" && zi.Spec.Zone != t.zone {
+			continue
+		}
+		if availableServicesEqual(availableServices, zi.Spec.GetAvailableServices()) {
+			continue
+		}
+		zi.Spec.AvailableServices = availableServices
+		if err := t.resManager.Update(ctx, zi); err != nil {
+			t.logger.Error(err, "couldn't update ZoneIngress", "name", zi.GetMeta().GetName())
```

Comment:
should we add something like "It will be updated on next reconciliation"? So when user read logs they can see error but then "update etc. etc."